proposals/0000-no-fallible-do.rst

```diff
@@ -0,0 +1,112 @@
+
+It's also unclear in general whether we want the unfolding of syntax sugar to depend on anything that is going on at the type level in the first place, as that makes it far more challenging to explain and understand.
+
+On the other hand, there are additionally cases where completeness checking of pattern synonyms becomes quite challenging as well, and ``{-# COMPLETE #-}`` pragmas may presently be required to convince GHC that certain pattern matches are complete. Even if those are specified, at present, the information about which ``COMPLETE`` pragmas exist is not available at the time of desugaring ``do``-syntax. Both making GHC take ``COMPLETE`` pragmas into consideration when deciding how to desugar `do`-expressions, and making further improvements to completeness checking in the face of pattern synonyms seem like good ideas, but there's certainly more work to be done in sorting out how this will actually be achieved in the context of GHC.
+
+These problems are frustrating in context, because often one has no interest in having ``MonadFail`` constraints appear at all, so one might want to avoid going down a rabbit hole of subtler issues of completeness checking when the real concern is for ``fail`` not to appear in the translation at all.
+
+So to cut through the knots a bit and provide a way to avoid the more complicated and challenging issues about completeness checking while their solutions are worked out, we decided to try something simple and provide a way to turn off the use of ``fail`` altogether.
+
+Proposed Change Specification
+-----------------------------
+
+We propose a module-level means of switching off the use of ``fail`` in ``do``-syntax altogether via an extension flag. Specifically, there is a default extension flag ``FallibleDo`` which indicates the usual translation of the ``do``-syntax involving ``fail``, and ``NoFallibleDo`` then replaces the use of ``fail`` with ``error``.
+
+Potentially failing pattern matches in the ``pat <- stmt`` syntax then result in a warning and the failing case is handled an a similar fashion to a failing pattern match in a lambda, with a generated application of ``error`` that provides the source location of the pattern match failure with a message.
+
+Examples
+--------
+
+If we take the example from the Motivation section:
+
+::
+
+  example x = do
+    A n <- x
+    return n
+
+With FallibleDo turned on (the default), this presently translates to:
+
+::
+
+  example x = x >>= \v -> case v of
+    A n -> return n
+    _ -> fail "..."
+
+which has an inferred type which is constrained by ``MonadFail m``
+
+with NoFallibleDo, this would become:
+
+::
+
+  example x = x >>= \v -> case v of
+    A n -> return n
+    _ -> error "..."
```

Comment:
Throw `PatternMatchFail` fail here instead, however one does that?

Reply:
throw (PatterMatchFail ...)